{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "22380059-712e-4643-90e3-7e6b845dbe88",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Importaciones"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "a925573f-6645-42ad-991c-cdef72691099",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import boto3\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import configparser\n",
    "import psycopg2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "59f4eb93-988c-487e-8b98-baf996826cd7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['config_db.cfg']"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config = configparser.ConfigParser()\n",
    "config.read('config_db.cfg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a63ed55c-a295-4fe5-bac8-35408c9462de",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'AKIAQ3EGVG3A2AT3OT56'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "config.get('IAM', 'ACCESS_KEY')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f36d58d3-d8f7-4505-a6f2-e4067997a57f",
   "metadata": {},
   "source": [
    "### Conexion con RDS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3f973783-c0b3-4326-899e-7f436d4d2b9e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "aws_rds_conn = boto3.client('rds', aws_access_key_id=config.get('IAM', 'ACCESS_KEY'),\n",
    "                           aws_secret_access_key=config.get('IAM', 'SECRET_KEY'),\n",
    "                           region_name='us-east-2')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "3bf1f375-1491-4a5e-9410-62cbe33f783d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Instancias disponibles: ['store-db-pg']\n"
     ]
    }
   ],
   "source": [
    "rds_instances_ids = []\n",
    "aws_response = aws_rds_conn.describe_db_instances()\n",
    "\n",
    "for response in aws_response['DBInstances']:\n",
    "    rds_instances_ids.append(response['DBInstanceIdentifier'])\n",
    "    \n",
    "print(f\"Instancias disponibles: {rds_instances_ids}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e405e8ba-89e2-4ad9-98fa-59ef1bab4f80",
   "metadata": {},
   "source": [
    "### Creacion de la Instancia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cbf041b5-3b28-4462-adfc-ce7f08ae684f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "La instancia ya existe\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    response = aws_rds_conn.create_db_instance(\n",
    "        DBInstanceIdentifier=config.get('DW', 'DB_INSTANCE_ID'),\n",
    "        DBName=config.get('DW', 'DB_NAME'),\n",
    "        MasterUsername=config.get('DW', 'DB_USERNAME'),\n",
    "        MasterUserPassword=config.get('DW', 'DB_PASSWORD'),\n",
    "        Port=int(config.get('DW', 'DB_PORT')),\n",
    "        DBInstanceClass='db.t3.micro',\n",
    "        Engine='mysql',\n",
    "        PubliclyAccessible=True,\n",
    "        AllocatedStorage=20,\n",
    "        VpcSecurityGroupIds=[config.get('VPC', 'SECURITY_GROUP')]\n",
    "    )\n",
    "    print('Creada exitosamente')\n",
    "except aws_rds_conn.exceptions.DBInstanceAlreadyExistsFault: \n",
    "    print(\"La instancia ya existe\")\n",
    "except Exception as ex:\n",
    "    print(\"Error\", ex)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "072f0a0f-07ef-4e9e-b2e0-d5d286ecfa23",
   "metadata": {},
   "source": [
    "### Obtenemos el nombre de la instancia"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e52af0f4-4859-41ba-b372-62cb25fa911f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dw-db-ms.c38kiu6mcwlc.us-east-2.rds.amazonaws.com\n"
     ]
    }
   ],
   "source": [
    "try: \n",
    "    instance = aws_rds_conn.describe_db_instances(DBInstanceIdentifier=config.get('DW', 'DB_INSTANCE_ID'))\n",
    "\n",
    "    RDS_HOSTNAME = instance.get('DBInstances')[0].get('Endpoint').get('Address')\n",
    "    print(RDS_HOSTNAME)\n",
    "except Exception as ex:\n",
    "    print(\"Error!\", ex)\n",
    "                                "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33449d7b-19e2-42f8-a436-7c53b0e65b20",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Librerias para MYSQL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "4da91240-7d32-4622-8e34-fdaec91d9019",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import DLL_DW_STR\n",
    "import mysql.connector as mysqlC"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f884d17-4390-4419-9951-8f608949595c",
   "metadata": {},
   "source": [
    "### Crear la base de datos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "c110e33c-f84d-4686-ac61-0506edba2e5f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: Error al crear la base de datos.\n",
      "1007 (HY000): Can't create database 'datawarehouse'; database exists\n"
     ]
    }
   ],
   "source": [
    "\n",
    "try:\n",
    "  schema_connection = mysqlC.connect(\n",
    "      host=RDS_HOSTNAME,\n",
    "      user=config.get('DW', 'DB_USERNAME'),\n",
    "      password=config.get('DW', 'DB_PASSWORD')\n",
    "  )\n",
    "  schema_cursor = schema_connection.cursor()\n",
    "  schema_cursor.execute(f\"CREATE DATABASE {config.get('DW', 'DB_NAME')}\")\n",
    "  schema_connection.commit()\n",
    "  print(\"Base de datos del Data Warehouse creada exitosamente\")\n",
    "except Exception as ex:\n",
    "  print(\"ERROR: Error al crear la base de datos.\")\n",
    "  print(ex)\n",
    "finally:\n",
    "  if schema_connection:\n",
    "    schema_connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5dbcbc29-d5bc-496b-b992-24f67b398ad9",
   "metadata": {},
   "source": [
    "### Verificar Conexion"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "e9c74f7a-5e30-4149-9a59-31803beb6529",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Conexion exitosa\n",
      "8.0.35\n",
      "Conectado a la base de datos: ('datawarehouse',)\n"
     ]
    }
   ],
   "source": [
    "\n",
    "try:\n",
    "    myDw = mysqlC.connect(\n",
    "        host=RDS_HOSTNAME,\n",
    "        user=config.get('DW', 'DB_USERNAME'),\n",
    "        password=config.get('DW', 'DB_PASSWORD'),\n",
    "        database=config.get('DW', 'DB_NAME')\n",
    "    )\n",
    "    \n",
    "    if myDw.is_connected():\n",
    "        print(\"Conexion exitosa\")\n",
    "        info_server=myDw.get_server_info()\n",
    "        print(info_server)\n",
    "        cursor=myDw.cursor()\n",
    "        print(\"Conectado a la base de datos: {}\".format(row))\n",
    "except Exception as ex:\n",
    "  print(\"ERROR: Error al crear el esquema del Data Warehouse.\")\n",
    "  print(ex)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "263b920f-a7d0-438c-8d18-18802960897b",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Crear el esquema"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "3e2a5fc9-7aa1-427b-a517-059775f5df82",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ERROR: Error al crear el esquema del Data Warehouse.\n",
      "Commands out of sync; you can't run this command now\n"
     ]
    },
    {
     "ename": "MySQLInterfaceError",
     "evalue": "Commands out of sync; you can't run this command now",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mMySQLInterfaceError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[60], line 17\u001b[0m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# Commit despuÃ©s de procesar las consultas DDL\u001b[39;00m\n\u001b[0;32m---> 17\u001b[0m myDw\u001b[38;5;241m.\u001b[39mcommit()\n\u001b[1;32m     19\u001b[0m \u001b[38;5;66;03m# Close the connection\u001b[39;00m\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/mysql/connector/connection_cext.py:532\u001b[0m, in \u001b[0;36mCMySQLConnection.commit\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    531\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandle_unread_result()\n\u001b[0;32m--> 532\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cmysql\u001b[38;5;241m.\u001b[39mcommit()\n",
      "\u001b[0;31mMySQLInterfaceError\u001b[0m: Commands out of sync; you can't run this command now",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mMySQLInterfaceError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[60], line 26\u001b[0m\n\u001b[1;32m     24\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mERROR: Error al crear el esquema del Data Warehouse.\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m     25\u001b[0m \u001b[38;5;28mprint\u001b[39m(ex)\n\u001b[0;32m---> 26\u001b[0m myDw\u001b[38;5;241m.\u001b[39mrollback()\n",
      "File \u001b[0;32m~/anaconda3/lib/python3.11/site-packages/mysql/connector/connection_cext.py:538\u001b[0m, in \u001b[0;36mCMySQLConnection.rollback\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    536\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cmysql:\n\u001b[1;32m    537\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cmysql\u001b[38;5;241m.\u001b[39mconsume_result()\n\u001b[0;32m--> 538\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_cmysql\u001b[38;5;241m.\u001b[39mrollback()\n",
      "\u001b[0;31mMySQLInterfaceError\u001b[0m: Commands out of sync; you can't run this command now"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    myDw = mysqlC.connect(\n",
    "        host=RDS_HOSTNAME,\n",
    "        user=config.get('DW', 'DB_USERNAME'),\n",
    "        password=config.get('DW', 'DB_PASSWORD'),\n",
    "        database=config.get('DW', 'DB_NAME')\n",
    "    )\n",
    "    mycursor = myDw.cursor()\n",
    "\n",
    "    # Execute the table creation statements from the script\n",
    "    mycursor.execute(DLL_DW_ST.DDL_QUERY)\n",
    "\n",
    "    # Procesar completamente los resultados antes de continuar\n",
    "    mycursor.fetchall()\n",
    "\n",
    "    # Commit despuÃ©s de procesar las consultas DDL\n",
    "    myDw.commit()\n",
    "\n",
    "    # Close the connection\n",
    "    if myDw:\n",
    "        myDw.close()\n",
    "    print(\"Esquema del Data Warehouse creado exitosamente\")\n",
    "except Exception as ex:\n",
    "    print(\"ERROR: Error al crear el esquema del Data Warehouse.\")\n",
    "    print(ex)\n",
    "    myDw.rollback()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9397dba6-afd4-4094-86b8-971abf6fcd01",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
